### PR TITLE
fix: PCRE lookahead-in-capturing-group and optional-quantifier backtrack (#36)

### LINKED ISSUE #36: [pcre] Lookahead combined with nested alternation produces wrong group captures
## Summary

Two PCRE tests involving lookahead assertions nested inside alternations or combined with digit-range character classes produce incorrect group captures.

## Failing PCRE Tests

1. Pattern `(\.\d\d((?=0)|\d(?=\d)))` on input `1.875000282`
   - Inner `(?=0)` / `\d(?=\d)` alternation inside a capturing group fails to record the correct group 2 value.

2. Pattern `(\.\d\d[1-9]?)\d+` on input `1.235`
   - Expected group 1 = `.23`, actual = `.235`
   - The `[1-9]?` optional class greedily consumes one character that should be left to `\d+`.

**Expected gain**: +2 PCRE conformance tests (Category 6, remaining after Phase 2.1)

## Root Cause

These are backtracking/greedy edge cases in patterns where a lookahead sits inside an alternation within a capturing group. The NFA/DFA grouping boundary isn't preserved correctly during lookahead evaluation and the greedy quantifier does not backtrack into the optional class.

## Implementation Notes

- Difficulty: Medium
- Files likely involved: `NFABytecodeGenerator.java`, lookahead handling in `PatternAnalyzer.java`

### PR DESCRIPTION
**Pull request by jbachorik**
### What does this PR do?

Fix two PCRE conformance bugs where lookahead assertions and optional greedy quantifiers inside capturing groups produce wrong results.

**Bug 1** — Pattern `(\.\d\d((?=0)|\d(?=\d)))` on `1.875000282`: group(2) returned the wrong value because the DFA strategy calls `match(substring)` for group extraction, truncating the lookahead context at the group boundary. Fix: detect lookahead-inside-capturing-group at analysis time and route to `OPTIMIZED_NFA_WITH_LOOKAROUND`.

**Bug 2** — Pattern `(\.\d\d[1-9]?)\d+` on `1.235`: group(1) returned `.235` instead of `.23` because the optional `[1-9]?` consumed `5` without backtracking when the continuation `\d+` needed it. Fix: extend `getGreedyGroupCharSet` to unwrap `ConcatNode` children and detect trailing optional quantifiers; teach the bytecode generator to emit two-path backtracking for those groups.

### Motivation

Conformance with the PCRE test suite — these were two remaining failures in category 6 after Phase 2.1.

### Related Issue(s)

Fixes #36

### Change Type

- [x] Bug fix
- [ ] New feature
- [ ] Performance improvement
- [ ] Refactoring (no functional change)
- [ ] Documentation
- [x] Test improvement
- [ ] Build/CI change

### Checklist

- [x] I have read the [CONTRIBUTING.md](../CONTRIBUTING.md) guidelines
- [x] All existing tests pass (`./gradlew build`)
- [x] I have added tests for my changes
- [ ] I have updated documentation (if applicable)
- [x] My commits are signed

### Performance Impact

No runtime impact — the new `hasLookaheadInsideCapturingGroup` check runs once per pattern compilation, not per match. The two-path backtracking path was already present for backref-trailing groups; extending it to quantifier-trailing groups adds no new code paths at match time.

### Additional Notes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by jbachorik** [comment]
@codex review

File: reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/analysis/PatternAnalyzer.java [hunk 1]
```diff
@@ -965,16 +1016,33 @@ private CharSet getGreedyGroupCharSet(RegexNode node) {
...
+    if (group.child instanceof ConcatNode) {
+      ConcatNode concat = (ConcatNode) group.child;
+      if (concat.children.isEmpty()) {
+        return null;
+      }
+      RegexNode last = concat.children.get(concat.children.size() - 1);
+      if (last instanceof QuantifierNode) {
+        QuantifierNode quant = (QuantifierNode) last;
+        if (quant.min == 0 && quant.max > 0 && quant.greedy) {
+          return getNodeCharSet(quant.child);
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/analysis/PatternAnalyzer.java]: Addressed — tightened the trailing-optional detection in getGreedyGroupCharSet to max == 1 so it stays in sync with hasTrailingOptionalQuantifier. The generator only handles the ? (max==1) case for two-path backtracking; {0,2} and similar bounded repeats are now correctly excluded from this path.

File: reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/analysis/PatternAnalyzer.java [hunk 2]
```diff
@@ -965,16 +1016,33 @@ private CharSet getGreedyGroupCharSet(RegexNode node) {
...
+        }
+      }
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/analysis/PatternAnalyzer.java:1044]: Addressed — same fix as above: tightened trailing-optional detection to max == 1 so the analyzer and generator stay aligned. Bounded repeats like {0,2} are now excluded from the two-path backtracking path.

File: reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/analysis/PatternAnalyzer.java [hunk 3]
```diff
@@ -413,7 +413,19 @@ public MatchingStrategyResult analyzeAndRecommend(boolean ignoreGroupCount) {
...
+        // Success! Assertions are simple literals, use DFA.
+        // However, when a lookahead appears inside a capturing group, the DFA path
+        // uses match(substring) for group extraction which breaks lookaheads that
+        // need to inspect characters beyond the match boundary. Route to NFA instead.
+        if (!ignoreGroupCount && hasLookaheadInsideCapturingGroup(ast)) {
+          return new MatchingStrategyResult(
+              MatchingStrategy.OPTIMIZED_NFA_WITH_LOOKAROUND,
+              null,
+              null,
+              false,
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/analysis/PatternAnalyzer.java:425]: Addressed — removed the !ignoreGroupCount guard so patterns with a lookahead inside a capturing group are routed to OPTIMIZED_NFA_WITH_LOOKAROUND even when called from hybrid mode (analyzeAndRecommend(true)). The RuntimeCompiler dfa==null check then falls back to OPTIMIZED_NFA, avoiding the truncated-substring NFA issue.

File: reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/codegen/RecursiveDescentBytecodeGenerator.java [hunk 4]
```diff
@@ -2233,13 +2257,13 @@ private void generateConcatWithBacktracking(ConcatNode node, int backtrackChildI
...
         // Generate two-path backtracking: try full group first, then mandatory-only.
         if (greedyChild instanceof GroupNode) {
           GroupNode backtrackGroup = (GroupNode) greedyChild;
           QuantifierNode trailingOpt = extractTrailingOptionalBackref(backtrackGroup.child);
-          if (trailingOpt != null) {
+          boolean needsTwoPathBacktrack =
+              trailingOpt != null || hasTrailingOptionalQuantifier(backtrackGroup.child);
+          if (needsTwoPathBacktrack) {
             // Emit a local fail label that returns -1
             Label localFail = new Label();
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/codegen/RecursiveDescentBytecodeGenerator.java:2268]: Addressed — extended generateRemainingWithBacktracking to also detect capturing groups with a trailing optional quantifier via hasTrailingOptionalQuantifier, matching the broader condition already used in generateConcatWithBacktracking. Two-path backtracking is now applied consistently across all remaining concat children.

**Comment by jbachorik** [comment]
@codex review

**Comment by jbachorik** [comment]
@Copilot review

File: reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/codegen/RecursiveDescentBytecodeGenerator.java [hunk 5]
```diff
@@ -2079,12 +2079,36 @@ private boolean containsBacktrackingQuantifier(RegexNode node) {
           if (extractTrailingOptionalBackref(g.child) != null) {
             return true;
           }
+          // Trailing optional greedy quantifier changes the group boundary, requiring
+          // the same two-path backtracking as a trailing optional backref.
+          if (hasTrailingOptionalQuantifier(g.child)) {
+            return true;
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/codegen/RecursiveDescentBytecodeGenerator.java:2085]: Addressed — extended generateNestedBacktracking to apply two-path backtracking for GroupNodes with a trailing optional quantifier. When the nestedQuantNode == null path encounters such a group, it now saves pos and groups into the free slotBase+0..1 slots, calls generateGroupWithOptionalBacktracking to try the full group followed by the mandatory-only fallback, and jumps to the outer backtrack loop only if both attempts fail.

**Comment by jbachorik** [comment]
@codex review

File: reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/analysis/PatternAnalyzer.java [hunk 6]
```diff
@@ -965,16 +1016,33 @@ private CharSet getGreedyGroupCharSet(RegexNode node) {
...
+    if (group.child instanceof ConcatNode) {
+      ConcatNode concat = (ConcatNode) group.child;
+      if (concat.children.isEmpty()) {
+        return null;
+      }
+      RegexNode last = concat.children.get(concat.children.size() - 1);
+      if (last instanceof QuantifierNode) {
+        QuantifierNode quant = (QuantifierNode) last;
+        if (quant.min == 0 && quant.max == 1 && quant.greedy) {
+          return getNodeCharSet(quant.child);
```
[the remark this answers is not in the document]
Comment (jbachorik) [reply reggie-codegen/src/main/java/com/datadoghq/reggie/codegen/analysis/PatternAnalyzer.java:1042]: Addressed — added getSuffixFirstCharSetSkippingNullable which accumulates first-char sets through consecutive nullable nodes (min==0 quantifiers, fully-nullable groups/concats). The overlap check now uses this helper instead of inspecting only the immediately next child, so patterns like ([1-9]?)-?\d+ are correctly routed to RECURSIVE_DESCENT because [1-9] is seen to overlap with \d+ once the nullable -? separator is looked through.
